rs/orchestrator/src/upgrade.rs

```diff
@@ -136,20 +140,69 @@ impl Upgrade {
     pub(crate) async fn check(&mut self) -> OrchestratorResult<Option<SubnetId>> {
         let latest_registry_version = self.registry.get_latest_version();
         // Determine the subnet_id using the local CUP.
-        let (subnet_id, local_cup_proto, local_cup) =
-            if let Some(proto) = self.cup_provider.get_local_cup_proto() {
-                let cup = CatchUpPackage::try_from(&proto).expect("deserializing CUP failed");
-                let subnet_id =
-                    get_subnet_id(&*self.registry.registry_client, &cup).map_err(|err| {
-                        OrchestratorError::UpgradeError(format!(
-                            "Couldn't determine the subnet id: {:?}",
-                            err
-                        ))
-                    })?;
-                (subnet_id, Some(proto), Some(cup))
-            } else {
-                // No local CUP found, check registry
-                match self.registry.get_subnet_id(latest_registry_version) {
+        let (subnet_id, local_cup_proto, local_cup) = {
+            let maybe_proto = self.cup_provider.get_local_cup_proto();
+            let maybe_cup = maybe_proto.as_ref().and_then(|proto| {
+                CatchUpPackage::try_from(proto)
+                    .map_err(|err| {
+                        error!(self.logger, "Failed to deserialize CatchUpPackage: {}", err);
+                    })
+                    .ok()
+            });
+
+            match (&maybe_cup, &maybe_proto) {
+                (Some(cup), _) => {
+                    let subnet_id =
+                        get_subnet_id(&*self.registry.registry_client, cup).map_err(|err| {
+                            OrchestratorError::UpgradeError(format!(
+                                "Couldn't determine the subnet id: {:?}",
+                                err
+                            ))
+                        })?;
+                    (subnet_id, maybe_proto, maybe_cup)
+                }
+                (None, Some(proto)) => {
+                    // We found a local CUP proto that we can't deserialize. This may only happen
+                    // if this is the first CUP we are reading on a new replica version after an
+                    // upgrade. This means we have to be an assigned node, otherwise we would have
+                    // left the subnet and deleted the CUP before upgrading to this version.
+                    // The only way to leave this branch is via subnet recovery.
```

Comment:
should we maybe log a critical error (and page on-calls) if we are in this branch?

Reply:
True, but we might get paged faster (although I don't know how long after a critical error counter is increased a page is emitted) and we would immediately know the reason for failed upgrade

Reply:
Done